instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/OperationListener.java

```diff
@@ -30,8 +30,9 @@ public interface OperationListener {
   /**
    * Listener method that is called at the end of an instrumented operation.
    *
+   * @param startAndEndAttributes The same object as startAttributes from onStart.
    * @param endNanos The nanosecond timestamp marking the end of the operation. Can be used to
    *     compute the duration of the entire operation.
    */
-  void onEnd(Context context, Attributes endAttributes, long endNanos);
+  void onEnd(Context context, Attributes startAndEndAttributes, long startNanos, long endNanos);
```

Comment:
Unfortunately `instrumentation-api`, that contains this class, is marked as stable. This means that backwards incompatible changes to API can not be made without bumping the major version. If we wish to proceed with this PR we need to figure out how to introduce these changes without breaking the API. @trask do you have any suggestions? We could duplicate the functionality into a new interface or I guess we could try something like
```java
default void onEnd(Context context, Attributes startAndEndAttributes, long startNanos, long endNanos) {
}

@Deprecated
default void onEnd(Context context, Attributes endAttributes, long endNanos) {
      throw new UnsupportedOperationException("This method is deprecated and will be removed in the next major release.");
}
```
usually we'd do that and make the new method delegate to the old, but here the attributes they expect are different. How would we tell which method to call? Use reflection on the target class and try to figure out which one it has? Add one more method to indicate which to call? Or indicate which to call when the operation listener is registered?

Reply:
Actually, typing that out made me wonder if something as simple as 
```
   default void onEnd(... startNanos...) {
      this.onEnd(originalVersion, noStartNanos); 
   }
```
But then overriding the new `onEnd` variation in all the in-repo implementations would work?

Reply:
Argh.  No, actually, any external implementations of this api might already be hurt by the assumption that `endAttributes` is *only* the end attributes, and not all of them.  So, there's probably no way around some sort of interface change, which could be done in a backward-compatible way (with extra logic and allocations to satisfy any such implementations).

Reply:
What if retaining `State` to pass the start timestamp in `onStart`, do not modify anything in `onStart` but let `startAttributes` be empty. By this way, it might be possible to avoid modifying the interface of `onEnd`

Reply:
@wgy035 That would still be a semantic change which might/would break repo-external implementations; the only options I see are bumping the semvar a major version (I wouldn't do that for this minor optimization, but only as part of a cohesive set of changes to minimize overhead) or to do the "v1/v2" interface concept with appropriate switching behavior in the usage of them.

Reply:
> Besides this awkward "v1/v2 interface,"

this seems ok to me